### PR TITLE
Provide callback to retrieve the SaveSubscriptionRequest to save link information

### PR DESCRIPTION
**Pull request by sohanirao**

### CODE REVIEW

File: src/runtime/link-accounts-flow-test.js [hunk 1]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
       expect(result).to.be.true;
     });
   });
+
+  it('should should fail if both token and authCode are present', () => {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: Just one "should"

File: src/runtime/link-accounts-flow-test.js [hunk 2]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+      throw new Error('must have failed');
+    }, reason => {
+      expect(() => {
+        throw reason;
+      }).to.throw(/Both authCode and token are available/);
+      return Promise.resolve();
+    });
+  });
+
+  it('should should fail if neither token nor authCode is present', () => {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: one "should"

File: src/runtime/link-accounts-flow-test.js [hunk 3]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+    activitiesMock.expects('openIframe').returns(Promise.resolve(port));
+    return linkSaveFlow.start().then(() => {
+      messageCallback({
+        'getLinkingInfo': true,
+      });
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      expect(messageStub).to.be.calledOnce.calledWith(
+        {authCode: 'testCode'});
+      return Promise.resolve();
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: remove

File: src/runtime/link-accounts-flow-test.js [hunk 4]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+    activitiesMock.expects('openIframe').returns(Promise.resolve(port));
+    return linkSaveFlow.start().then(() => {
+      messageCallback({
+        'getLinkingInfo': true,
+      });
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      expect(messageStub).to.be.calledOnce.calledWith(
+        {token: 'test'});
+      return Promise.resolve();
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: remove

File: src/runtime/link-accounts-flow-test.js [hunk 5]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+        'getLinkingInfo': true,
+      });
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      throw new Error('must have failed');
+    }, reason => {
+      expect(() => {
+        throw reason;
+      }).to.throw(/Both authCode and token are available/);
+      return Promise.resolve();
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: remove

File: src/runtime/link-accounts-flow-test.js [hunk 6]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+    activitiesMock.expects('openIframe').returns(Promise.resolve(port));
+    return linkSaveFlow.start().then(() => {
+      messageCallback({
+        'getLinkingInfo': true,
+      });
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      throw new Error('must have failed');
+    }, reason => {
+      expect(() => {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: At least one of failure tests should check that the dialog is closed.

File: src/runtime/link-accounts-flow-test.js [hunk 7]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+    return linkSaveFlow.start().then(() => {
+      messageCallback({
+        'getLinkingInfo': true,
+      });
+      dialogManagerMock.expects('completeView').once();
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      throw new Error('must have failed');
+    }, reason => {
+      expect(() => {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: This test also needs to check that the dialog has been closed.

File: src/runtime/link-accounts-flow-test.js [hunk 8]
```diff
@@ -470,4 +440,117 @@ describes.realWin('LinkSaveFlow', {}, env => {
...
+      return linkSaveFlow.requestPromise_;
+    }).then(() => {
+      expect(messageStub).to.be.calledOnce.calledWith(
+        {authCode: 'testCode'});
+      return Promise.resolve();
+    });
+  });
+
+  it('should callback promise rejected should close dialog', () => {
+    linkSaveFlow = new LinkSaveFlow(runtime, () => Promise.reject('no token'));
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: We need a similar test where the callback throws synchronously. E.g. `() => {throw new Error('failed');}`.

File: src/runtime/link-accounts-flow.js [hunk 9]
```diff
@@ -184,16 +184,23 @@ export class LinkCompleteFlow {
   }
 }
 
+/**
+ * Callback for retrieving subscription request
+ *
+ * @callback saveRequestCallback
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: TIL. Looks good, but let's capitalize.

File: src/runtime/link-accounts-flow.js [hunk 10]
```diff
@@ -206,8 +213,11 @@ export class LinkSaveFlow {
     /** @private @const {!../components/dialog-manager.DialogManager} */
     this.dialogManager_ = deps.dialogManager();
 
-    /** @private {!../api/subscriptions.SaveSubscriptionRequest} */
-    this.request_ = request;
+    /** @private {!saveRequestCallback} */
+    this.callback_ = callback;
+
+    /** {?Promise<../api/subscriptions.SaveSubscriptionRequest>} */
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: `@private` and the `SaveSubscriptionRequest` must be `!`. And further, we is the `SaveSubscriptionRequest` in the `@callback` definition is used w/o namespace and with namespace here. Does compiler give any warnings about this?

File: src/runtime/link-accounts-flow.js [hunk 11]
```diff
@@ -223,25 +233,44 @@ export class LinkSaveFlow {
...
       this.win_,
       this.activityPorts_,
       feUrl('/linksaveiframe'),
       feArgs(iframeArgs),
       /* shouldFadeBody */ false
     );
+
+    this.activityIframeView_.onMessage(data => {
+      if (data['getLinkingInfo']) {
+        /** {!Promise<../api/subscriptions.SaveSubscriptionRequest>} */
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: remove

File: src/runtime/link-accounts-flow.js [hunk 12]
```diff
@@ -223,25 +233,44 @@ export class LinkSaveFlow {
...
+        this.requestPromise_ = new Promise(resolve => {
+          resolve(this.callback_());
+        }).then(request => {
+          if (!request) {
+            throw new Error('Neither token or authCode is available');
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: Duplicated from below. You can do it this way:

```
if (request && request.token) {
} else if (request && request.authCode)  {
} else {
  throw new Error('Both ...');
}
```

File: src/runtime/link-accounts-flow.js [hunk 13]
```diff
@@ -184,16 +184,23 @@ export class LinkCompleteFlow {
...
+
 /**
  * The flow to save subscription information.
  */
 export class LinkSaveFlow {
 
   /**
    * @param {!./deps.DepsDef} deps
-   * @param {!../api/subscriptions.SaveSubscriptionRequest} request
+   * @param {saveRequestCallback} callback
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: `!`

File: src/runtime/link-accounts-flow.js [hunk 14]
```diff
@@ -184,16 +184,23 @@ export class LinkCompleteFlow {
...
    */
-  constructor(deps, request) {
+  constructor(deps, callback) {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: This means that you also need to similarly modify `saveSubscription()` API in the `subscriptions.js` to switch to callback.

File: src/runtime/link-accounts-flow-test.js (hunk 1 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 6 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done, discussed offline.

File: src/runtime/link-accounts-flow-test.js (hunk 5 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 2 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 4 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 3 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 8 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 7 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow-test.js]: done, discussed offline.

File: src/runtime/link-accounts-flow.js (hunk 9 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done.

File: src/runtime/link-accounts-flow.js (hunk 13 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done.

File: src/runtime/link-accounts-flow.js (hunk 10 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/link-accounts-flow.js (hunk 11 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/link-accounts-flow.js (hunk 12 above)
Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done

File: docs/link-flow.md [hunk 15]
```diff
@@ -28,13 +28,22 @@ SwG supports two flavors of OAuth account linking:
...
-A list of user's accounts will be surfaced and the user can select the account to be linked to the provided token. Upon saving this link, a non-blocking confirmation will be shown to the user.
+```
+const requestPromise = new Promise(resolve => {
+    // whenever available
+    resolve({authCode: 'THE CODE'});
+});
+subscriptions.saveSubscription(() => requestPromise));
+```
+
+A list of user's accounts will be surfaced in an iframe and the user can select the account to be linked and saved. At this point, the publisher will receive a message to retrieve the token or authCode using the callback. Once its available,  the iframe recieves it and saves it. Upon saving this link, a non-blocking confirmation will be shown to the user.
```
Comment (dvoytenko) [inline docs/link-flow.md]: A bit too much of implementation detail. Let's say something like: "The dialog will prompt the user to opt-in to save the subscription. If the user agrees, the provided callback will be called to resolve the token/authCode value. The resulting promise will be resolved once the subscription has been saved".

File: src/api/subscriptions.js [hunk 16]
```diff
@@ -231,6 +231,14 @@ export let LoginRequest;
  */
 export let SaveSubscriptionRequest;
 
+/**
+ * Callback for retrieving subscription request
+ *
+ * @callback SaveSubscriptionRequestCallback
+ * @return {!Promise<SaveSbuscriptionRequest> | !SaveSubscriptionRequest} request
+ */
+export let SaveSubscriptionRequestCallback;
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: `saveSubscription()` above still needs to be updated to use this type.

File: src/runtime/link-accounts-flow.js [hunk 17]
```diff
@@ -206,13 +206,20 @@ export class LinkSaveFlow {
...
+
+    /** @private {?Promise<!../api/subscriptions.SaveSubscriptionRequest>} */
+    this.requestPromise_ = null;
 
     /** @private {?ActivityIframeView} */
     this.activityIframeView_ = null;
   }
 
+  /** @return {?Promise<!../api/subscriptions.SaveSubscriptionRequest>} */
+  getRequestPromise() {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: Is it necessary to export it as a public API? Unfortunately currently all public APIs are preserved in the minified version, but it looks like this method is only needed for testing?

Comment (sohanirao) [reply src/runtime/link-accounts-flow.js]: done

File: src/api/subscriptions.js (hunk 16 above)
Comment (sohanirao) [reply src/api/subscriptions.js]: done

File: docs/link-flow.md (hunk 15 above)
Comment (sohanirao) [reply docs/link-flow.md]: done

**Pull request merged**